### PR TITLE
feat: Public API for CT Framework Definitions

### LINKED ISSUE #25638: Implement Public API for Component Testing Onboarding
### What would you like?

Based on the work done in the technical brief: https://github.com/cypress-io/cypress/issues/24991 (Cypress team only), implement the proposed public API.

Documentation: (public post 12.7 release): https://docs.cypress.io/guides/component-testing/third-party-definitions

### Why is this needed?

_No response_

### Other

_No response_

### PR DESCRIPTION
**Pull request by lmiller1990**
Closes https://github.com/cypress-io/cypress/issues/25638

**Note**: No merge yet- still a known issue, see end of the PR description.

Code was reviewed separately in

- https://github.com/cypress-io/cypress/pull/25713
- https://github.com/cypress-io/cypress/pull/25749

A second pass is always good, though! 🙏 

- Technical Brief: https://github.com/cypress-io/engineering-documentation/blob/main/tech-briefs/ct-public-api/README.md
- Example third party Component Framework Definition: https://github.com/lmiller1990/cypress-ct-solid-js

## Screenshots

(1) it adds a community definition with relevant tag in launchpad

(2) Known bug in dep detection - not introduced here - see #25782 

(3) it works!

## Test Plan

1. make Solid.js app: `npx degit solidjs/templates/js my-app`
2. `cd my-app`
3. Install deps and my simple [Solid.js definition](https://github.com/lmiller1990/cypress-ct-solid-js): `npm install cypress-ct-lachlan-solid-js`

Two options, either
1. clone this branch, `yarn install` so you have the latest everything, `yarn cypress:open` at monorepo root for global mode, choose `my-app` (the one you just made)
2. grab a binary: https://github.com/cypress-io/cypress/commit/b2f6343ab7ae088c3907f38acdc10a73aafc517b#comments

Then...

1. Go through onboarding - Solid should show up. Use Vite, the template in step 1 is a Vite one.
2. It won't properly detect that you installed Solid.js. Known issue, unrelated to this PR, see #25782. Just hit Skip
3. Should work like any other CT lib, example test you can use:
4. wasn't that easy!! almost the same as a first party onboarding :D

```js
import App from '../../src/App'

describe('App', () => {
  it('playground', () => {
    cy.mount(<App />)
    cy.get('a').contains('Learn Solid')
  })
})
```

## Known issues/TODO before merge

- [x] ~Not working with binary. Definitions fail to resolve with `Ignoring TypeError: Invalid host defined options`. Seems to be V8 related according to [this tweet](https://twitter.com/slicknet/status/1559619176230047744), which is the only info I can find about this error. Working on it.~ Edit seems to be fixed https://github.com/cypress-io/cypress/commit/b2f6343ab7ae088c3907f38acdc10a73aafc517b#comments
- [ ] Fix flaky test (launchpad test for adding Qwik, a third party definition)

## Fast follows

- docs https://github.com/cypress-io/cypress-documentation/issues/5039
- #25782
- show at Cypress App demo
- GTM plan 

These are all out of scope for this PR - but we will do those next sprint.

### CODE REVIEW

File: packages/data-context/src/actions/WizardActions.ts
Comment (lmiller1990) [inline packages/data-context/src/actions/WizardActions.ts:170]: Found a race condition, if this is slow, the UI is wrong

image
https://app.circleci.com/pipelines/github/cypress-io/cypress/49442/workflows/64f996ae-9b7e-4e74-b7d7-a091da0047c6/jobs/2054201/tests#failed-test-0

I will fix it

File: packages/launchpad/cypress/e2e/scaffold-component-testing.cy.ts
Comment (lmiller1990) [inline packages/launchpad/cypress/e2e/scaffold-component-testing.cy.ts:167]: There is a race condition causing CI to fail here. Introduced here, since getting frameworks is now an async operation, since we glob in `node_modules`.

We need to fix this before merging.

**Comment by astone123** [comment]
Any reason the Vite icon looks different between community and first-party frameworks?
Screenshot 2023-02-13 at 11 10 19 AM
Screenshot 2023-02-13 at 11 10 34 AM

**Comment by astone123** [comment]
I'm also not seeing the mount log in the command log... it looks like [in your adapter](https://github.com/lmiller1990/cypress-ct-solid-js/blob/main/index.mjs#L16) you call `Cypress.log` to print this out but it's not showing up for some reason

File: packages/launchpad/cypress/e2e/scaffold-component-testing.cy.ts
Comment (ZachJW34) [reply packages/launchpad/cypress/e2e/scaffold-component-testing.cy.ts:167]: I tried reproducing this and could not get it to flake. I tried locally as well as the circle runner (running 20+ times). I have an idea that might fix this. Maybe we can pair later on this

**Comment by ZachJW34** [comment]
@astone123 Looked into the Vite logo issue. Kind of convoluted but this is what is happening:

- we use `vite-svg-loader` to load svgs which has an optimization step. Something like `<linear-gradient id="my-unique-id">` gets transformed to `<linear-gradient id="a">` (minification)
- We load an external icon that has `<linear-gradient id="a">` defined, one of the gradients gets overwritten.

The solidjs svg icon we are loading is stomping on the vite logo since they both end up with the same id for the gradient. A fix for this is to make the external svg have unique ids, so rather than `a` it would be `solidjs-a`. We could also try and fix this within our Vite config by modifying the options we pass to `vite-svg-loader`. I tried this but ended up breaking most of our icons 😅 

I wouldn't say this is a blocker though

**Review by ZachJW34 — APPROVED** [review]
Looks good to me! We should hold off on merging this feature branch till after the release

**Comment by marktnoonan** [comment]
Yeah this was fine for me with no issues that weren't already known. Very happy to see it. Will hold off on approval till we see if there's more changes coming for the race condition https://github.com/cypress-io/cypress/pull/25780#discussion_r1104966416

**Comment by astone123** [comment]
Any idea why we're not seeing the log message in the command log when the component mounts? I was messing around and wrote an [adapter for Preact](https://github.com/astone123/cypress-ct-test-preact) and my log wasn't showing up either. Looks like this is the same thing that we do for first-class integrations, but it works there.

**Comment by lmiller1990** [comment]
@astone123 Maybe you need a default `options.log ?? true`?  Looking at your code, options defaults to `{}`. so when you do `if (Cypress.log)` it will always  be false.

Eg maybe: https://github.com/astone123/cypress-ct-test-preact/pull/1

Official adapters do `log !== false` (see [here](https://github.com/cypress-io/cypress/blob/2c355109383a4b570216a0e60e3ab26d328da29c/npm/vue/src/index.ts#L429) for example). I think you might need to do something similar.

**Review by astone123 — APPROVED** [review]

**Review by jordanpowell88 — COMMENTED** [review]
@lmiller1990  Is the `defineComponentFramework()` function not suppose to include the `icon` property? I noticed this when trying to use it to help with the qwik implementation

**Comment by lmiller1990** [comment]
> @lmiller1990 Is the `defineComponentFramework()` function not suppose to include the `icon` property? I noticed this when trying to use it to help with the qwik implementation

@jordanpowell88 Yep, just noticed this myself. Fixed https://github.com/cypress-io/cypress/pull/25780/commits/598be2db1aa5c7559c9590df5cc945f5fe1c8c15

**Comment by lmiller1990** [comment]
This is looking good, did final smoke test on windows, all good. Known CI failures are unrelated and fixed by: https://github.com/cypress-io/cypress/pull/25889

I am going to admin merge this as CI was green and the only change since was to resolve conflicts in `circleci/workflows.yml`.

**Comment by lmiller1990** [comment]
This is looking good, did final smoke test on windows, all good. Known CI failures are unrelated and fixed by: https://github.com/cypress-io/cypress/pull/25889

**Pull request merged**
